src/scipp/core/dimensions.py

```diff
@@ -0,0 +1,73 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+# @author Simon Heybrock
+from .._scipp.core import Variable, DataArray, Dataset, CoordError
+from .dataset import merge
+
+
+def _make_sizes(obj):
+    """
+    Makes a dictionary of dimensions labels to dimension sizes
+    """
+    return dict(zip(obj.dims, obj.shape))
+
+
+def _rename_variable(var: Variable, dims_dict: dict = None, **names) -> Variable:
+    """
+    Rename the dimensions labels of a Variable.
+    The renaming can be defined:
+       - using a dict mapping the old to new names, e.g. rename({'x': 'a', 'y': 'b'})
+       - using keyword arguments, e.g. rename(x='a', y='b')
+    """
+    return var.rename_dims({**({} if dims_dict is None else dims_dict), **names})
+
+
+def _rename_data_array(da: DataArray, dims_dict: dict = None, **names) -> DataArray:
+    """
+    Rename the dimensions, coordinates and attributes of a Dataset.
+    The renaming can be defined:
+       - using a dict mapping the old to new names, e.g. rename({'x': 'a', 'y': 'b'})
+       - using keyword arguments, e.g. rename(x='a', y='b')
+    """
+    renaming_dict = {**({} if dims_dict is None else dims_dict), **names}
+    out = da.rename_dims(renaming_dict)
+    coords_and_attrs = list(zip(("coord", "attr"), (out.coords, out.attrs)))
+    for old, new in renaming_dict.items():
+        for i, (group, meta) in enumerate(coords_and_attrs):
+            if old in meta:
+                if new in meta:
+                    raise CoordError(
+                        f"Cannot rename {group} '{old}' to '{new}' as this would erase "
+                        f"an existing {group}.")
+                elif new in out.meta:
+                    other = coords_and_attrs[i - 1][0]
+                    raise CoordError(
+                        f"Cannot rename {group} '{old}' to '{new}' as there is an "
+                        f"existing {other} with the same name.")
+                meta[new] = meta.pop(old)
```

Comment:
Four nested levels is a bit much. How about something like
```python
for old, new in renaming_dict.items():
  if new in da.meta:
    raise CoordError()
```
Also, isn't there an issue with popping/inserting from a dict during iteration?

Reply:
> Four nested levels is a bit much. How about something like
> 
> ```python
> for old, new in renaming_dict.items():
>   if new in da.meta:
>     raise CoordError()
> ```

So you would have a single, not quite as descriptive, error?

> Also, isn't there an issue with popping/inserting from a dict during iteration?

I don't think I am iterating over the dicts of coords or attrs, i am iterating over the `renaming_dict`, and then finding the coords I want in the coords and attrs.

Reply:
Right, I misread the `if` for a `for`.

Yes, I think it is fine to just have an error `"Cannot rename x to y since a coord or attr named y exists"`.